### PR TITLE
Improve error handling

### PR DESCRIPTION
**Pull request by larryonoff**
- `tasker/______.js` adds more error codes, this's a rough implemented feature.
- the most important changes are in `tasker/android.js`.
- `tasker/runServerRemote__.sh` container very controversial change. That I think should be removed now or later.

---
This change is [Reviewable](https://reviewable.io/reviews/thaliproject/ci/92)

### CODE REVIEW

**Comment by czyzm** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[tasker/runServerRemote__.sh, line 29 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_rJOs0fXl3FRzrb4n:-KZ_rJOtVaXkAxJoq5NM:b5gygc5) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/runServerRemote__.sh#L29)):*
> ```Shell
> # so we're skipping errors for 'sudo pkill jx'
> # MUST be fixed
> ! sudo pkill jx
> ```

It seems for me as a safety guard so that in case previous jx process is still active we kill it. If there is no jx process running then pkill jx will return exit code 1.
In other words returning exit code 1 here is fine. 
Of course pkill may return 1 also due to other errors and if we would like to be fancy we should check first with pgrep if process exists and just then kill it but I think for now it is enough to leave it as is.
Just add comment why we are ignoring the error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by czyzm** [comment]
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[tasker/android.js, line 94 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_uXSNhiAU_4ClXIU2:-KZ_uXSNhiAU_4ClXIU3:b-x8obbn) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/android.js#L94)):*
> ```JavaScript
>     if (err ||
>         stdout.indexOf(class_name) === -1 ||
>         stdout.indexOf('Success') === -1) {
> ```

I think we should introduce similar check in runAndroidApp. It also uses adb shell and is just checking exitCode which will be 0 even if app is not started. We could check for existence of 'Error' in the output.
Here is the successful run output:
```
Starting: Intent { cmp=com.rockwellautomation.AppPlatform/.MainActivity }`
```
and here the failure:
```
Starting: Intent { cmp=com.rockwellautomation.AppPlatform/.MainActivita }
Error type 3
Error: Activity class {com.rockwellautomation.AppPlatform/com.rockwellautomation.AppPlatform.MainActivita} does not exist.
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by czyzm** [comment]
Reviewed 1 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[tasker/______.js, line 28 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_vzrjpUO5MTLvzHkS:-KZ_vzrjpUO5MTLvzHkT:b-jt9ajj) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/______.js#L28)):*
> ```JavaScript
>   var message = errors[code-1];
>   if (message) {
>     console.error(errors[code-1], args);
> ```

You can use message instead of errors[code-1] 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by czyzm** [comment]
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[tasker/______.js, line 80 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_x2Ya59_Lso5jPHff:-KZ_x2Ya59_Lso5jPHfg:b-95zsgk) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/______.js#L80)):*
> ```JavaScript
> 
>     if (out.exitCode != 0) {
>       console.log(out.out, "\n");
> ```

No need to log it as it will be logged by logErrorAndExit

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by czyzm** [comment]
Reviewed 1 of 3 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[tasker/______.js, line 83 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_xKMuVVtAZm0mvVVa:-KZ_xKMuVVtAZm0mvVVb:b5cibzm) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/______.js#L83)):*
> ```JavaScript
>       // If jx install fails, we can't run the server
>       // and without the server, the tests wouldn't run
>       logErrorAndExit(3, [out.exitCode, out.out]);
> ```

The code should be 4,  not 3 (look at error messages)

---

*[tasker/______.js, line 106 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_xW8w_dLNp1csNk9T:-KZ_xW8w_dLNp1csNk9U:bkkxhjf) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/______.js#L106)):*
> ```JavaScript
> 
>       child.on('exit', function (code) {
>         logErrorAndExit(4, [code]);
> ```

And here should be 5.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by larryonoff** [comment]
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[tasker/______.js, line 28 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_vzrjpUO5MTLvzHkS:-KZaCCB0x9K4_V0m-aWK:b-o3hl83) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/______.js#L28)):*
<details><summary><i>Previously, czyzm wrote…</i></summary><blockquote>

You can use message instead of errors[code-1] 
</blockquote></details>

great catch. thank you!

---

*[tasker/______.js, line 80 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_x2Ya59_Lso5jPHff:-KZaCFgpzpZdMUKLx46T:b-896fix) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/______.js#L80)):*
<details><summary><i>Previously, czyzm wrote…</i></summary><blockquote>

No need to log it as it will be logged by logErrorAndExit
</blockquote></details>

Done.

---

*[tasker/______.js, line 83 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_xKMuVVtAZm0mvVVa:-KZaCG9kzV89i-pw0JJd:b5xt3rx) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/______.js#L83)):*
<details><summary><i>Previously, czyzm wrote…</i></summary><blockquote>

The code should be 4,  not 3 (look at error messages)
</blockquote></details>

thanks for catching. I've counted from 0. :(

---

*[tasker/android.js, line 94 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_uXSNhiAU_4ClXIU2:-KZaCZVNFA_B_3aYAUov:b-896fix) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/android.js#L94)):*
<details><summary><i>Previously, czyzm wrote…</i></summary><blockquote>

I think we should introduce similar check in runAndroidApp. It also uses adb shell and is just checking exitCode which will be 0 even if app is not started. We could check for existence of 'Error' in the output.
Here is the successful run output:
```
Starting: Intent { cmp=com.rockwellautomation.AppPlatform/.MainActivity }`
```
and here the failure:
```
Starting: Intent { cmp=com.rockwellautomation.AppPlatform/.MainActivita }
Error type 3
Error: Activity class {com.rockwellautomation.AppPlatform/com.rockwellautomation.AppPlatform.MainActivita} does not exist.
```

</blockquote></details>

Done.

---

*[tasker/runServerRemote__.sh, line 29 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_rJOs0fXl3FRzrb4n:-KZaCbgadNEepeXyuvq8:b-sss13d) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/runServerRemote__.sh#L29)):*
<details><summary><i>Previously, czyzm wrote…</i></summary><blockquote>

It seems for me as a safety guard so that in case previous jx process is still active we kill it. If there is no jx process running then pkill jx will return exit code 1.
In other words returning exit code 1 here is fine. 
Of course pkill may return 1 also due to other errors and if we would like to be fancy we should check first with pgrep if process exists and just then kill it but I think for now it is enough to leave it as is.
Just add comment why we are ignoring the error.
</blockquote></details>

I added check that verifies that js is running and in this case executes `pkill jx`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by larryonoff** [comment]
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[tasker/______.js, line 106 at r1](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZ_xW8w_dLNp1csNk9T:-KZaComWFiVN3GpukLRN:b-896fix) ([raw file](https://github.com/thaliproject/ci/blob/37dcf04d461ba7ebe0377861bbaba6ddab16fe6f/tasker/______.js#L106)):*
<details><summary><i>Previously, czyzm wrote…</i></summary><blockquote>

And here should be 5.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by czyzm** [comment]
Reviewed 2 of 3 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by czyzm** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92#-:-KZac4lV5jKluz-pxoZr:bnfp4nl)*

**Comment by jareksl** [comment]
*[tasker/______.js, line 17 at r3](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZacFlcZ_igHGn3tfq8:-KZacFlcZ_igHGn3tfq9:bvctioz) ([raw file](https://github.com/thaliproject/ci/blob/dd8d478981c43dd192d4c417aee72cee3ee8361b/tasker/______.js#L17)):*
> ```JavaScript
> };
> 
> var errors = [
> ```

NB You could also define errors as object literal, so later you could simply retrieve the message using just code instead of code -1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by jareksl** [comment]
:lgtm:

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92#-:-KZadRu5GCFjV_3ZEw9f:bnfp4nl)*

**Comment by jareksl** [comment]
Reviewed 2 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by larryonoff** [comment]
@yaronyg please review too

**Comment by yaronyg** [comment]
Please fix the index issue but otherwise :lgtm: 

---

Reviewed 2 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[tasker/______.js, line 17 at r3](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZacFlcZ_igHGn3tfq8:-KZbZ8G0T6xJJ721j19-:b-s4i3zf) ([raw file](https://github.com/thaliproject/ci/blob/dd8d478981c43dd192d4c417aee72cee3ee8361b/tasker/______.js#L17)):*
<details><summary><i>Previously, jareksl (Jarosław Słota) wrote…</i></summary><blockquote>

NB You could also define errors as object literal, so later you could simply retrieve the message using just code instead of code -1
</blockquote></details>

+1. Using numeric IDs is really hard to read.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92#-:-KZbZnkjhLVSZL0xQ0Tx:bbxcjs9)*

**Comment by larryonoff** [comment]
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[tasker/______.js, line 17 at r3](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZacFlcZ_igHGn3tfq8:-KZenLD875va1ABUQ6ve:b-jlgeq) ([raw file](https://github.com/thaliproject/ci/blob/dd8d478981c43dd192d4c417aee72cee3ee8361b/tasker/______.js#L17)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary><blockquote>

+1. Using numeric IDs is really hard to read.
</blockquote></details>

Please check if I did it properly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by jareksl** [comment]
*[tasker/______.js, line 97 at r4](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZfDKIdD8rIe0OEZ7wZ:-KZfDKIdD8rIe0OEZ7w_:b-co4nzf) ([raw file](https://github.com/thaliproject/ci/blob/77a5577f6cefb6743ba25459e94ca97884c1a92c/tasker/______.js#L97)):*
> ```JavaScript
>       // If jx install fails, we can't run the server
>       // and without the server, the tests wouldn't run
>       logAndExit('jxInstallFailed', [out.exitCode, out.out]);
> ```

Please note also that you have already object with all codes, so instead of passing string and comparing it in the logAndExit function, easier would be to just pass property like following:  logAndExit(errors.indexJsFailed, [out.exitCode, out.out]); 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by jareksl** [comment]
*[tasker/______.js, line 70 at r4](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZfHnkUE8pn3O06hz5H:-KZfHnkUE8pn3O06hz5I:bv4dona) ([raw file](https://github.com/thaliproject/ci/blob/77a5577f6cefb6743ba25459e94ca97884c1a92c/tasker/______.js#L70)):*
> ```JavaScript
>   targets.droid = 0;
> }
> targets.cancel=1;
> ```

Since you're editing the code, please add spaces around "=", just to be consistent

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by jareksl** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by larryonoff** [comment]
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[tasker/______.js, line 70 at r4](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZfHnkUE8pn3O06hz5H:-KZfOptiNOnKxA3gXGW4:b-896fix) ([raw file](https://github.com/thaliproject/ci/blob/77a5577f6cefb6743ba25459e94ca97884c1a92c/tasker/______.js#L70)):*
<details><summary><i>Previously, jareksl (Jarosław Słota) wrote…</i></summary><blockquote>

Since you're editing the code, please add spaces around "=", just to be consistent
</blockquote></details>

Done.

---

*[tasker/______.js, line 97 at r4](https://reviewable.io:443/reviews/thaliproject/ci/92#-KZfDKIdD8rIe0OEZ7wZ:-KZfOrvnkCeyE-qbPEQF:b-896fix) ([raw file](https://github.com/thaliproject/ci/blob/77a5577f6cefb6743ba25459e94ca97884c1a92c/tasker/______.js#L97)):*
<details><summary><i>Previously, jareksl (Jarosław Słota) wrote…</i></summary><blockquote>

Please note also that you have already object with all codes, so instead of passing string and comparing it in the logAndExit function, easier would be to just pass property like following:  logAndExit(errors.indexJsFailed, [out.exitCode, out.out]); 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*

**Comment by jareksl** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/ci/92)*
